packages/firestore/src/local/persistence.ts

```diff
@@ -29,14 +29,25 @@ import { RemoteDocumentCache } from './remote_document_cache';
 import { ClientId } from './shared_client_state';
 
 /**
- * Opaque interface representing a persistence transaction.
+ * A base class representing a persistence transaction, encapsulating both the
+ * transaction's sequence numbers as well as a list of onCommitted listeners.
  *
  * When you call Persistence.runTransaction(), it will create a transaction and
  * pass it to your callback. You then pass it to any method that operates
  * on persistence.
  */
 export abstract class PersistenceTransaction {
+  private readonly onCommittedListeners: Array<() => {}> = [];
+
   abstract readonly currentSequenceNumber: ListenSequenceNumber;
+
+  addOnCommittedListener(listener: () => {}): void {
```

Comment:
Shouldn't we require that the committed listeners return `void`? That way you can't accidentally have a return statement in your callback that you intend for the outer method.

Or would that interfere with using short arrow functions in the callback?

Reply:
I just realized this when I started to modify my "applyRemoteEvent" PR. It is much better if we return void.